ls1intum/Artemis
# `Programming exercises`: Add option to export results with test cases and feedback

### Checklist
#### General


#### Client
#### ~Changes affecting Programming Exercises~

### Motivation and Context


Some instructors suggested the integration of this feature in order to increase the flexibility provided by the exported CSV results file. 

### Description

This PR replaces, in the scores view of programming exercises, the export results button with a dropdown one containing three possible export formats. 

* **Without test cases**: the standard results export format as used before
* **With test cases**: additionally to the standard results, this includes a `passed` or `failed` information for each of the exercise's test cases. If information about the test case cannot be found (i.e. on failed builds), `skipped` is used.
* **With test cases and feedback**: this includes the same additional data as **With test cases** but includes a copy of the provided feedback in case of failed test cases

An example of the three formats can be seen below:

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise with scores
- 1 Text Exercise (or any exercise type except for programming ones)

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to the programming exercise's score view
4. Export the results using all three formats and check its resulting files
5. Navigate to the "non-programming" exercise
6. Export its scores and check the resulting file


### Review Progress






#### Code Review
#### Manual Tests
#### ~Exam Mode Test~

### Test Coverage




| Class/File | Branch | Line |
|------------|-------:|-----:|
| exercise-scores-export-button.component.ts | 77.65% | 96.87% |


### Screenshots



![image](https://user-images.githubusercontent.com/45761921/205661900-ed432156-b49b-4ac1-a420-034ad7a5a314.png)


**Without test cases**:

![image](https://user-images.githubusercontent.com/45761921/205673664-d45296b3-18a4-4221-a299-2dab2b307c16.png)


**With test cases**: 

![image](https://user-images.githubusercontent.com/45761921/205673513-a5a9ece8-c830-4058-90b1-22643271417e.png)

**With test cases and feedback**:

![image](https://user-images.githubusercontent.com/45761921/205674086-b75fee3c-f1a6-4b86-aca9-e4fce9628d52.png)
